### PR TITLE
Sp2 single tack

### PR DESCRIPTION
**Pull request by dan-rds**
Sprint 2, User story 2: (Daniel's portion)
- RouteModel datatype
- Cart_Point datatype
- Basic single tack route calculation (with wind taken into account)
- Documentation on how to use the lib (see README)

### CODE REVIEW

File: lib/calc_route.dart [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+  var results = parser.parse(args);
+  v_flag = results['verbose'];
+
+  Cart_Point start = new Cart_Point(0.0, 0.0);
+  Cart_Point end = new Cart_Point(1.0, -10.2);
+  double wind = 3 * pi / 2;
+  RouteModel route = new RouteModel(start: start, end: end, wind_radians: wind);
+
+  calculateOptimal(route);
+  print(route);
```
Comment (donovanrost) [inline lib/calc_route.dart:23]: Should we have print statements in here?

File: lib/calc_route.dart [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+  RouteWriter rw = new RouteWriter();
+  rw.writeToFileFromRouteModel(route: route);
+  rw.run_python_plotter();
+}
+
+double euc_dist(Cart_Point a, Cart_Point b) {
```
Comment (donovanrost) [inline lib/calc_route.dart]: https://www.dartlang.org/guides/language/effective-dart/style I think lowerCamelCase for etc_dist

File: lib/calc_route.dart [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  return a.toPoint().distanceTo(b.toPoint());
+}
+
+void verbose_print(dynamic s) {
```
Comment (donovanrost) [inline lib/calc_route.dart]: https://www.dartlang.org/guides/language/effective-dart/style I think lowerCamelCase for verbose_print

File: lib/calc_route.dart [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  if (v_flag) {
+    print(s);
+  }
+}
+
+Cart_Point find_intersection(Cart_Point a_point, Cart_Point b_point, Cart_Point c_point, Cart_Point d_point) {
```
Comment (donovanrost) [inline lib/calc_route.dart]: https://www.dartlang.org/guides/language/effective-dart/style I think lowerCamelCase for find_intersection

File: lib/calc_route.dart [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    double x = (b2 * c1 - b1 * c2) / determinant;
+    double y = (a1 * c2 - a2 * c1) / determinant;
+
+    verbose_print("intersection: " + Cart_Point(x, y).toString());
+
+    return new Cart_Point(x, y);
+  }
+}
+
+bool check_viable_tack(Cart_Point p1, Cart_Point p2, double wind_radians) {
```
Comment (donovanrost) [inline lib/calc_route.dart]: https://www.dartlang.org/guides/language/effective-dart/style I think lowerCamelCase for check_viable_tack

File: lib/calc_route.dart [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+  /*
+    check_viable_tack just checks if a tack from p1 -> p2 is makable (not directly upwind).
+    Args:
+        p1 (Cart_Point): Cartesian coordinates of start
+        p2 (Cart_Point): Cartesian coordinates of end
+        wind_radians (double): wind direction 
+    Returns:
+        bool
+  */
+  double no_go_limit_radians = 0.610865; // Equivilant to 35 degrees
```
Comment (donovanrost) [inline lib/calc_route.dart:74]: https://www.dartlang.org/guides/language/effective-dart/style I think lowerCamelCase for no_go_limit_radians

File: lib/calc_route.dart [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+  var angle_from_ideal_lower = (impossible_tack - ngzone_lower_bound);
+
+  verbose_print("possible tacks: " + radToDegStr(angle_from_ideal_lower) + " <-> " + radToDegStr(angle_from_ideal_upper));
+  bool lower_better = (angle_from_ideal_lower < angle_from_ideal_upper);
+  lower_better = (invert_flag) ? !lower_better : lower_better;
+
+  return (lower_better) ? ngzone_lower_bound : ngzone_upper_bound;
+}
+
+calculateOptimal(RouteModel route) {
```
Comment (donovanrost) [inline lib/calc_route.dart:113]: calculate optimal WHAT?

File: lib/calc_route.dart [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+  List intermediate_points = new List<Cart_Point>();
+  bool viable = check_viable_tack(route.start, route.end, route.wind_radians);
+  if (!viable) {
+    double best_tack_slope = bestTackAngle(route.start, route.end, route.wind_radians, false);
```
Comment (donovanrost) [inline lib/calc_route.dart:117]: https://www.dartlang.org/guides/language/effective-dart/style I think lowerCamelCase for best_tack_slope

File: lib/calc_route.dart [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+    verbose_print("__BEST TACK " + radToDegStr(best_tack_slope));
+    // TODO this can be done better, note th +pi
+    double inverse_tack_slope = bestTackAngle(route.end, route.start, route.wind_radians, true) + pi;
```
Comment (donovanrost) [inline lib/calc_route.dart:121]: https://www.dartlang.org/guides/language/effective-dart/style I think lowerCamelCase for inverse_tack_slope

File: lib/cart_point.dart [hunk 10]
```diff
@@ -0,0 +1,29 @@
+import 'dart:math';
+
+class Cart_Point {
```
Comment (donovanrost) [inline lib/cart_point.dart]: https://www.dartlang.org/guides/language/effective-dart/style I think UpperCamelCase for Cart_Point

File: lib/route_model.dart [hunk 11]
```diff
@@ -0,0 +1,55 @@
+import 'dart:core';
+import 'dart:math';
+
+import 'package:sail_routing_dart/cart_point.dart';
+
+class RouteModel {
+  Cart_Point _start;
+  Cart_Point _end;
+  List<Cart_Point> _intermediate_points;
+  double _wind_radians;
```
Comment (donovanrost) [inline lib/route_model.dart:9]: https://www.dartlang.org/guides/language/effective-dart/style I think lowerCamelCase for _wind_radians

File: lib/route_model.dart [hunk 12]
```diff
@@ -0,0 +1,55 @@
+import 'dart:core';
+import 'dart:math';
+
+import 'package:sail_routing_dart/cart_point.dart';
+
+class RouteModel {
+  Cart_Point _start;
+  Cart_Point _end;
+  List<Cart_Point> _intermediate_points;
```
Comment (donovanrost) [inline lib/route_model.dart]: https://www.dartlang.org/guides/language/effective-dart/style I think lowerCamelCase for _intermetiate_points

File: lib/route_model.dart [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+  double _wind_radians;
+
+  // Getters
+  Cart_Point get start => this._start;
+  Cart_Point get end => this._end;
+  List<Cart_Point> get intermediate_points => this._intermediate_points;
+  double get wind_radians => this._wind_radians;
```
Comment (donovanrost) [inline lib/route_model.dart:15]: https://www.dartlang.org/guides/language/effective-dart/style I think lowerCamelCase for wind_radians

File: lib/route_model.dart [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+  void set intermediate_points(List<Cart_Point> l) =>
```
Comment (donovanrost) [inline lib/route_model.dart]: https://www.dartlang.org/guides/language/effective-dart/style I think lowerCamelCase for intermediate_points

File: lib/calc_route.dart (hunk 1 above)
Comment (dan-rds) [reply lib/calc_route.dart:23]: Yes, that is a testing script.

File: lib/route_model.dart [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+    output += "\n[/ROUTE]";
+    return output;
+  }
+
+  calculateOptimal() {}
+}
+
+void main() {
+  Cart_Point p = new Cart_Point(1.0, 1.0);
+  print(p);
```
Comment (donovanrost) [inline lib/route_model.dart]: should we still have print statements in here?

File: lib/calc_route.dart (hunk 3 above)
Comment (dan-rds) [reply lib/calc_route.dart]: good p

File: lib/route_model.dart (hunk 15 above)
Comment (dan-rds) [reply lib/route_model.dart]: Fixed

File: lib/calc_route.dart (hunk 7 above)
Comment (dan-rds) [reply lib/calc_route.dart:113]: Lol the only arg is `route`

File: lib/calc_route.dart [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+  return a.toPoint().distanceTo(b.toPoint());
+}
+
+void verbosePrint(dynamic s) {
+  if (v_flag) {
+    print(s);
+  }
+}
+
+CartPoint find_intersection(CartPoint a_point, CartPoint b_point, CartPoint c_point, CartPoint d_point) {
```
Comment (nkalscheuer) [inline lib/calc_route.dart:40]: We should make this a vector2 function, shouldn't be much of a change from what you have already though. Also, this should be a part of a shared lib file. I added the dart file. We could talk more about it tomorrow

File: lib/cart_point.dart [hunk 17]
```diff
@@ -0,0 +1,29 @@
+import 'dart:math';
```
Comment (nkalscheuer) [inline lib/cart_point.dart:1]: We can probably drop this class now that we have the vector library

File: lib/route_writer.dart [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+  writeToFileFromRouteModel({RouteModel route}) {
+    List fileContents = new List();
+    fileContents.add({'start': route.start.toList()});
+
+    fileContents.add({'end': route.end.toList()});
+    fileContents.add({'wind_direction': route.wind_radians});
+
+    fileContents.add({'points': route.intermediate_points.map((CartPoint p) => p.toList()).toList()});
+
+    var myFile = new File('lib/plotter_data.json');
```
Comment (nkalscheuer) [inline lib/route_writer.dart:34]: This is going to be a problem running dart other than from main, can we look into using some sort of package path?

Comment (nkalscheuer) [reply lib/route_writer.dart:34]: Or be able to pass the file path in as an arg?

**Review by nkalscheuer — COMMENTED** [review]
Looks good, but I wasn't able to get a newly generated graph, maybe we can look into it tomorrow after the TA meeting.

**Pull request merged**